src/librustc/middle/dead.rs

```diff
@@ -231,6 +247,10 @@ impl<'a, 'tcx> Visitor<'tcx> for MarkSymbolVisitor<'a, 'tcx> {
             hir::ExprTupField(ref lhs, idx) => {
                 self.handle_tup_field_access(&lhs, idx.node);
             }
+            hir::ExprStruct(ref qpath, ref fields, _) => {
+                let def = self.tables.qpath_def(qpath, expr.id);
```

Comment:
The last thing is to make it work for type aliases (it'll also fix the ICE with `Self`).

You can use `self.tcx.expr_ty(expr)` to get the struct expression's type and then check that it's a union.

Test:
```rust
union U {
    a: u8, // should not be reported
    b: u8, // should not be reported
    c: u8, // should be reported
}
type A = U;

let u = A { a: 0 };
let b = unsafe { u.b };
```